undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRemove.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.undertow.filters;
+
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.ServiceRemoveStepHandler;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceName;
+
+/**
+ * @author baranowb
+ */
+public class FilterRemove extends ServiceRemoveStepHandler {
+
+    public FilterRemove(AbstractAddStepHandler addOperation) {
+        super(addOperation);
+    }
+
+    public FilterRemove(ServiceName baseServiceName, AbstractAddStepHandler addOperation) {
+        super(baseServiceName, addOperation);
+    }
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+            throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+
+        context.deregisterCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition().getDynamicName(context.getCurrentAddress()));
+    }
+
+}
```

Comment:
You do not need to subclass ServiceRemoveStepHandler.
Capability deregistration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractRemoveStepHandler.java#L175

Reply:
Same with this one, explicitly requested by Brian.

Reply:
Provided that the filter's ResourceDefinition registered the RuntimeCapability (and it should), the capability will be auto-deregistered with the OperationContext by the super implementation.